src/lib.rs

```diff
@@ -1,12 +1,1184 @@
+                            PtpContainerType::Data,
+                            code,
+                            self.current_tid,
+                            data);
+            let timespec = time::get_time();
+            trace!("Write Data [{}:{:09}] - 0x{:04x} ({}), tid:{}, len:{}",
+                   timespec.sec,
+                   timespec.nsec,
+                   code,
+                   StandardCommandCode::name(code).unwrap_or("unknown"),
+                   self.current_tid,
+                   data.len());
+            try!(self.handle.write_bulk(self.ep_out, &data_message, timeout));
+        }
+
+        self.current_tid += 1;
+
+        let mut data = None;
+        loop {
+            unsafe {
+                self.buf.set_len(0);
+            }
+
+            loop {
+                let chunk_size = 256 * 1024;
+                let current_len = self.buf.len();
+                let current_capacity = self.buf.capacity();
+                if current_capacity - current_len < chunk_size {
+                    self.buf.reserve(chunk_size);
+                }
+                let remaining_buf = unsafe {
+                    ::std::slice::from_raw_parts_mut(self.buf.get_unchecked_mut(current_len) as *mut _, chunk_size)
+                };
+                let timespec = time::get_time();
+                trace!("Read Data  [{}:{:09}] - length:{:?} remaining:{:?}",
+                       timespec.sec,
+                       timespec.nsec,
+                       current_len,
+                       remaining_buf.len());
+                let len = try!(self.handle.read_bulk(self.ep_in, remaining_buf, timeout));
+                unsafe {
+                    self.buf.set_len(current_len + len);
+                }
+                // debug!("new buf len [{:?}] into {:?}", self.buf.len(), remaining_buf.len());
+                if len < remaining_buf.len() {
+                    break;
+                }
+            }
+
+            let (mtype, mut msg) = try!(PtpTransaction::parse(&self.buf));
+
+            if mtype == PtpContainerType::Data && msg.is_response(&transaction) {
+                data = Some(msg.data);
+            } else if mtype == PtpContainerType::Response && msg.is_response(&transaction) {
+                if let Some(data) = data {
+                    msg.data = data;
+                }
+                
+                if msg.code != StandardResponseCode::Ok {
+                    return Err(Error::Response(msg.code));
+                }
+                
+                return Ok(msg.data);
+            }
+        }
+    }
+
+    pub fn get_objectinfo(&mut self, handle: u32) -> Result<PtpObjectInfo, Error> {
+        let data = try!(self.command(StandardCommandCode::GetObjectInfo, &vec![handle], None));
+        Ok(try!(PtpObjectInfo::decode(&data)))
+    }
+
+    pub fn get_object(&mut self, handle: u32) -> Result<Vec<u8>, Error> {
+        // TODO why need this loop?
```

Comment:
Let's remove this loop. If it's needed in our other lib we'll implement the logic there.